COOKBOOK.md

```diff
@@ -0,0 +1,145 @@
+
+Iterators avoid making those intermediate allocations by executing all their specified transformations on an individual item before moving on to the next item.
+
+Using iter-tools, the above example would look like:
+```js
+import { map, filter, compose, toArray } from 'iter-tools';
+
+const activeUserNames = compose(
+    toArray,
+    filter(name => name.length > 0),
+    map(([id, user]) => [user.first, user.last].join(' ')),
+    filter(([id, user]) => user.status === 'active'),
+    filter(Boolean),
+  )(users);
+```
+
+## The toolbox that should have come with [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).
+
+The silliest thing about es6 Maps is that they don't come with any tooling.
+
+None of the standard functional utilites like `map` or `filter` can be found on `Map.prototype`. Instead there are only a few iterators exposed, and a constructor that takes iterables... My personal theory is that the language authors recognized that these methods could be easily (and reusably) implemented by an iterator library.
+
+So here they are, the missing methods for Maps!
+
+```js
+const users = new Map([
+  [1, {name: 'Ada Lovelace', status: 'active'}],
+  [3, {name: null, status: 'banned'}],
+]);
+
+const posts = new Map([
+  [1, {user: 1, text: 'Hello world!'}],
+  [2, {user: 1, text: 'The only certain things in life are death and taxes.'}],
+  [3, {user: 3, text: 'HUGE SALES AT CARPET EMPORIUM! SAVE SAVE SAVE!'}]
+]);
+
+const postsByActiveUsers = new Map(compose(
+  filter(([id, post]) => post.user.status === 'active'),
+  map(([id, post]) => ({...post, user: users.get(post.user.id)})),
+)(users));
+```
+
+## Reusing Entries
+The iter-tools `entries` function takes a second parameter, `reuseEntry`. 
+
+This parameter is strictly a performance optimization. In the case of small objects its impact will be minimal. **It should not be used if you are not sure that it is safe!** Use the examples below as guidance as to what is safe.
+
+Reusing entries avoids needless memory allocation. In each of the examples below, passing true for `reuseEntry` will eliminate the need to allocate an entire set of entry objects, saving the garbage collector work later.
+
+This is because it is imperative that Maps never expose their internal entries tuples, whose mutation would trigger real state changes to (or inconsistencies in) the data structure. Thus a full set of entries objects is normally allocated even when the usage, as in these examples, does not strictly require it.
+
+### Instantiating Maps
+```js
+const myMap = new Map(entries({
+  foo: 'foo',
+  bar: 'bar',
+}, true));
+```
+
+### Destructured iteration on Maps
+```js
+for (const [key, value] of entries(myMap, true)) {
```

Comment:
```js
for (const [key, value] of myMap) { ...
```
would be the same ...

Reply:
Ah, almost the same, but not the same. The `true` parameter is causing entries to be reused meaning there's no allocations happening on each iteration. The same is not true of a Map's default iterator. That's the point of this bit of the cookbook!

I'll definitely add the caveat that the default behavior is fine in most situations.